File: C++/323.number-of-connected Components-in-an-undirected-graph.cpp
/*
 * @lc app=leetcode id=323 lang=cpp
 *
 * [323] Number of Connected Components in an Undirected Graph
 *
 * Given n nodes labeled from 0 to n - 1 and a list of undirected edges (each edge is a pair of nodes), write a function to find the number of connected components in an undirected graph.
 * 
 * Example 1:
 * 
 *     0          3
 *
 *     |          |
 * 
 *     1 --- 2    4
 * 
 * 
 * Given n = 5 and edges = [[0, 1], [1, 2], [3, 4]], return 2.
 * 
 * Example 2:
 * 
 *    0           4
 *
 *    |           |
 *
 *    1 --- 2 --- 3
 * 
 * Given n = 5 and edges = [[0, 1], [1, 2], [2, 3], [3, 4]], return 1.
 * 
 * Note:
 * 
 * You can assume that no duplicate edges will appear in edges. Since all edges are undirected, [0, 1] is the same as [1, 0] and thus will not appear together in edges.
 * 
 * 
 */
#include<vector>
using namespace std;

// @lc code=start
class Solution {
public:
    // int countComponents(int n, vector<pair<int, int> >& edges) {
    //     int ans = 0;
    //     vector<vector<int>> adj(n);
    //     for (auto edge: edges) {
    //         adj[edge.first].push_back(edge.second);
    //         adj[edge.second].push_back(edge.first);
    //     }

    //     vector<bool> visited;
    //     for (int i=0; i<n; ++i) {
    //         if (!visited[i]) {
    //             DFS(adj, visited, i);
    //             ans ++;
    //         }
    //     }
    //     return ans;
    // }
    // void DFS(vector<vector<int>>& adj, vector<bool>& visited, int i) {
    //     if (visited[i]) return;
    //     visited[i] = true;
    //     for (int j=0; j<adj[i].size(); ++j) {
    //         DFS(adj, visited, adj[i][j]);
    //     }
    // }
    int countComponents(int n, vector<pair<int, int> >& edges) {
        int ans = n;
        vector<int> root(n, -1);
        // for (int i=0; i<n; ++i) root[i] = i;
        for (auto edge: edges) {
            int p1 = findRoot(root, edge.first);
            int p2 = findRoot(root, edge.second);
            if (p1 != p2) {
                root[p1] =  p2;
                ans --;
            }
        }
        return ans;
    }
    int findRoot(vector<int>& root, int i) {
        // while (root[i] != i) i = root[i];
        while (root[i] != -1) i = root[i];
        return i;
    }
};
// @lc code=end

File: Python/323.number-of-connected-components-in-an-undirected-graph.py
#
# @lc app=leetcode id=323 lang=python3
#
# [323] Number of Connected Components in an Undirected Graph
#
# https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/description/
#
# algorithms
# Medium (62.85%)
# Likes:    2662
# Dislikes: 101
# Total Accepted:    408.5K
# Total Submissions: 648.5K
# Testcase Example:  '5\n[[0,1],[1,2],[3,4]]'
#
# You have a graph of n nodes. You are given an integer n and an array edges
# where edges[i] = [ai, bi] indicates that there is an edge between ai and bi
# in the graph.
# 
# Return the number of connected components in the graph.
# 
# 
# Example 1:
# 
# 
# Input: n = 5, edges = [[0,1],[1,2],[3,4]]
# Output: 2
# 
# 
# Example 2:
# 
# 
# Input: n = 5, edges = [[0,1],[1,2],[2,3],[3,4]]
# Output: 1
# 
# 
# 
# Constraints:
# 
# 
# 1 <= n <= 2000
# 1 <= edges.length <= 5000
# edges[i].length == 2
# 0 <= ai <= bi < n
# ai != bi
# There are no repeated edges.
# 
# 
#

# @lc code=start
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        ans = n

        root = [i for i in range(n)]
        for e in edges:
            p1 = self.findRoot(root, e[0])
            p2 = self.findRoot(root, e[1])
            if p1 != p2:
                root[p1] = p2
                ans -= 1
        return ans
        
    
    def findRoot(self, root, i):
        while root[i] != i:
            root[i] = root[root[i]]
            i = root[i]
        return i
# @lc code=end

